### DISCUSSION TITLE
Generate knotvector for interpolation

### DISCUSSION THREAD

**Discussion by rendetto** [nortikin/sverchok#4804] [Q&A] [answered]
From https://github.com/nortikin/sverchok/pull/4789 description:

> This node provides several ways of generating a knotvector for building **(or interpolation)** of NURBS curves and surfaces.

@portnov Can you please shed some light how can we use the generated KV for interpolation? I cannot find a node that needs knotvector input and does interpolation of some sort. For sure I'm missing something.

Thanks

**Comment by portnov** [comment, marked as answer]
In Sverchok-Extra, there is a set of nodes for Nurbs curve interpolation and approximation:
![Screenshot_20221208_175629](https://user-images.githubusercontent.com/284644/206452278-9a5be309-2e2d-468d-8d9a-08c92a863e06.png)
I have no idea yet how useful they are in practice, so for now they are only in Sverchok-Extra.
They are based on the following idea: if you want to find a curve, which goes through the given point, you in fact have one equation. If you want this curve go through N given points, then you have N equations. Also you may want to specify not only points through which the curve goes, but some other conditions, for example, curve tangent vectors at some points.
So, if you want this curve to have M control points, and you have N conditions of some sort, you have a system of equations with M unknowns and N equations.
If M is equal to N, than this system is fully-determined, and you can solve it and obtain a single solution - a curve which satisfies all conditions. This is a case of interpolation.
If M < N, then the system is overdetermined, and so, in general, it does not have exact solutions. But we can use least squares method to find an approximate solution: a curve which approximately satisfies all conditions. This is the case of approximation.
And it is possible that M > N, i.e. the system is underdetermined. In this case, there is, in general, an infinite number of curves which satisfy all conditions. We cannot generate them all. But, in some tasks, in addition, we have a given ("initial") curve with M control points, and we want to slightly move it's control points so that it would satisfy our conditions. Then, we can use the same principle of least squares, to find the curve which differs from our initial curve as least as possible, but does satisfy conditions.
All the above assumes that for each point you have in your conditions, you have corresponding T value on the curve. So you have to either explicitly provide the knotvector of the curve to be found (this is here "generate knotvector" node can be useful), or try to calculate it based on locations of the points, based on some metric.
The workflow of these nodes is similar to one used in Pulga Physics. You set up several nodes which generate "Goals" (I don't want to call them conditions, because in case of overdetermined system they will not be satisfied precisely), then you combine them in a list, and pass to "Solver" node, which solves the equations and outputs the curve.

**Reply by rendetto** [reply]
Thanks for the detailed explanation. This is a lot to digest :) I was suspecting those new nodes but I don't have much success utilizing them ...yet.

**Reply by portnov** [reply]
@rendetto An example of interpolation with these nodes:
![Screenshot_20221208_182020](https://user-images.githubusercontent.com/284644/206456863-d1034830-36b5-4d80-be83-2481375f36b5.png)
An example of use of underdetermined system - modify the curve so that it would go through three specified points:
![Screenshot_20221208_181611](https://user-images.githubusercontent.com/284644/206456940-57db4c0a-9cc6-4f1a-af7e-b1480d7a30a8.png)
